source/request_source/request_source_plugin_impl.h

```diff
@@ -0,0 +1,124 @@
+// Implementations of RequestSourceConfigFactory and the RequestSources that those factories make.
+#pragma once
+
+#include "envoy/registry/registry.h"
+
+#include "nighthawk/common/request_source_plugin_config_factory.h"
+
+#include "external/envoy/source/common/common/lock_guard.h"
+#include "external/envoy/source/common/common/thread.h"
+
+#include "api/client/options.pb.h"
+#include "api/request_source/request_source_plugin.pb.h"
+
+#include "common/uri_impl.h"
+
+namespace Nighthawk {
+
+// Stub Request Source implementation for comparison.
+class DummyRequestSource : public RequestSource {
+public:
+  explicit DummyRequestSource(const nighthawk::request_source::StubPluginConfig& config);
+  // The generator function will only return empty headers.
+  // The function is threadsafe.
+  RequestGenerator get() override;
+
+  // default implementation
+  void initOnThread() override;
+};
+
+// Factory that creates a DummyRequestSource from a DummyRequestSourcePluginConfig proto.
+// Registered as an Envoy plugin.
+// Stub implementation of RequestSourceConfigFactory which produces a RequestSource.
+// RequestSources are used to get RequestGenerators which generate requests for the benchmark
+// client. All plugins configuration are specified in the request_source_plugin.proto This class is
+// thread-safe, but it doesn't do anything. Usage: assume you are passed an appropriate Any type
+// object called config, an Api object called api, and a default header called header. auto&
+// config_factory =
+//     Envoy::Config::Utility::getAndCheckFactoryByName<RequestSourcePluginConfigFactory>(
+//         "nighthawk.stub-request-source-plugin");
+// RequestSourcePtr plugin =
+//     config_factory.createRequestSourcePlugin(config, std::move(api), std::move(header));
+
+class DummyRequestSourcePluginConfigFactory : public virtual RequestSourcePluginConfigFactory {
+public:
+  // This is a hardcoded string.
+  std::string name() const override;
+  // This returns an empty version of the expected StubPluginConfig from request_source_plugin.proto
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+  // This is the primary method that is used to get RequestSources.
+  // This implementation is thread safe, but the RequestSource it generates doesn't do much.
+  RequestSourcePtr createRequestSourcePlugin(const Envoy::Protobuf::Message& message,
+                                             Envoy::Api::ApiPtr api,
+                                             Envoy::Http::RequestHeaderMapPtr header) override;
+};
+
+// This factory will be activated through RequestSourceFactory in factories.h
+DECLARE_FACTORY(DummyRequestSourcePluginConfigFactory);
+
+// Sample Request Source for small RequestOptionsLists. Loads a copy of the RequestOptionsList in
+// memory and replays them.
+// @param request_max The number of requests the requestGenerator produced by get() will generate. 0
```

Comment:
If I understand correctly what "request_max" specifies, a name like "total_requests" would probably be more descriptive.

Reply:
Agreed.